Implementation of CAP-0034 for stellar-protocol issue #622: Preserve Transaction-Set/Close-Time Affinity During Nomination

rokopt opened the pull request:
# Description

Implements [CAP-0034](https://github.com/stellar/stellar-protocol/pull/679), and thereby resolves [stellar-protocol issue #622](https://github.com/stellar/stellar-protocol/issues/622).

For details, see that PR and issue, and the commit logs for the "State pattern" and "Upgrade protocol" commits in this PR.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -697,11 +700,20 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
 
     // take the txSet with the biggest size, highest xored hash that we have
     TxSetFramePtr bestTxSet;
+    std::vector<TimePoint> candidateCloseTimes;
```
Why not change `candidateValues` to be a `vector<StellarValue, ValueWrapperPtrSet::iterator>` ?
Then you don't need to track all those things separately, and when picking the best tx set, the index is an iterator in that vector (instead of using optional indexes, etc).

rokopt replied:
I think I've found something even simpler:  just track the maximum `closeTime` _and_ the `closeTime` and signature associated with the best tx set.  Then at the end you just choose once (based on the protocol version) whether it's the max or the best-associated `closeTime` that you take.  That way, there's no need for any new vectors at all, or any iterators or indexes.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
                 {
                     highestTxSet = cTxSet;
                     highest = sv.txSetHash;
+                    if (candidateCloseTimes.size() == 0)
```
this cannot happen, a few lines up you're inserting into the vector

rokopt replied:
Alright, another `logic_error` bites the dust.  :)

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
+                    }
+                    highestTxSetIndex =
+                        make_optional<size_t>(candidateCloseTimes.size() - 1);
+                    if (candidateCloseTimes[*highestTxSetIndex] != sv.closeTime)
```
this becomes irrelevant if you use one data structure instead of many

rokopt replied:
It's been removed, although a different way (see comment above).

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
             }
         }
+
+        if (!highestTxSet)
```
no need for this check: this would be a bug internal to that method

rokopt replied:
Removed.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
 
+    std::vector<TransactionFrameBasePtr> removed;
+    if (!highestTxSetIndex)
+    {
```
same thing: this is just writing code that attempts to verify itself for correctness - we just end up with confusing code

rokopt replied:
Alright, removed.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -711,22 +723,82 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
+    // This ->trimInvalid() should be a no-op:  we should have trimmed any
+    // expired transactions before nominating a candidate, and should have
+    // validated that other nodes have done so.  We call it as an assertion.
+    removed = bestTxSet->trimInvalid(
```
this code is confusing and the only thing that can come out of it is a crash: in protocol 14 we don't modify the value, so there is no need to verify that it's still valid (check itself can be wrong)

rokopt replied:
Alright, I've removed it.  Since nothing ever sets `removed` now before checking its `size()`, I've also removed the whole  `if (size() != 0)` block below.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -324,39 +342,29 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
         res = SCPDriver::kInvalidValue;
     }
-    else if (!txSet->checkValid(mApp, 0))
+    else if (!txSet->checkValid(mApp, closeTimeOffsetToTrimOrValidateAgainst(
```
I don't think this is what CAP0034 says: we expect txsets to all be valid for the new close time.

This change implies that we expect them to be valid from `lcl.ct` up to `new_ct` (inclusive) which is a lot more restrictive.

rokopt replied:
I had failed to glean that that meant to check over a range and not an offset.  I guess the difference is in whether it returns `txTOO_EARLY` or not; is that right, or are there any other potential differences?

rokopt replied:
And actually, don't we check a transaction for validity as soon as we receive it, using the last ledger header close time (in particular to check for `txTOO_EARLY`), in `TransactionQueue::canAdd()`?  And when a ledger closes, we trim transactions against the new last ledger close time, in `updateTransactionQueue()`, correct?

Consequently, won't we already have validated a transaction against the last ledger header close time before we reach the quoted check, so that checking for `txTOO_EARLY` against the last ledger header close time is harmless (redundant), and won't happen, and the only semantic effect of the change is the intended CAP-34 effect of checking for `txTOO_LATE` against a later `closeTime`?

rokopt replied:
> And actually, don't we check a transaction for validity as soon as we receive it, using the last ledger header close time (in particular to check for `txTOO_EARLY`), in `TransactionQueue::canAdd()`? And when a ledger closes, we trim transactions against the new last ledger close time, in `updateTransactionQueue()`, correct?

Upon even further thought, I think that my second sentence, about `updateTransactionQueue()`, is unnecessary.  If a transaction in the queue were going to be `txTOO_EARLY` at the time of a ledger close, then it would have been `txTOO_EARLY` at the time of any _previous_ ledger close, too, so it would have been rejected before even entering the queue, in `TransactionQueue::canAdd()`.  So the validation in `TransactionQueue::canAdd()` has, I believe, already ensured that any transaction that makes it to `validateValueHelper()` (above) or `triggerNextLedger()` (below) _is_ valid with respect to `lcl.ct`.  Validating again against `lcl.ct` (as well as `new_ct`) at those points is therefore not any more restrictive than validating against that whole range.

rokopt replied:
> Validating again against lcl.ct (as well as new_ct) at those points is therefore not any more restrictive than validating against that whole range.

Except that, on yet _more_ consideration, I realize that that's only assuming that you're well-behaved!  When we validate a `StellarValue` in `triggerNextLedger()` or `validateValueHelper()`, it might have come from an ill-behaved node.  So the question is, what's the correct behavior if the `StellarValue` came from an ill-behaved node which had not validated against `lcl.ct` when it accepted the transaction into its queue -- to check `txTOO_EARLY` the way my code does, and as the code before CAP-0034 already did, against `lcl.ct`, or the way you suggest, against `new_ct`?

I have not, so far, been able to think of an attack that a bad validator could use to take advantage of the weaker validation that you suggest, but in general it seems to me that if the runtime cost is the same, then we should validate as strongly as we believe we should be able to.  Put another way:  if some other node did something that we ourselves _wouldn't_ have done in the same situation, then I think we probably shouldn't trust it.

MonsieurNicolas replied:
You're confusing what is defined at the application protocol level from the rest:
* at the protocol level, we have very strict set of validity checks. Basically anything that happens during consensus. Those rules have to be as minimal as possible as to avoid tying up changes in overlay to protocol changes.
* outside of the protocol, we have other things that we can do but that are not mandated at the network level. Typically those things are there to make our life easier and can change at any time, or just the result that individual nodes on the network are not trusted. This includes:
    * whatever heuristic we want to use at the overlay layer.
    * the transaction set composition (ie, how to fit as many good transactions in a transaction set)

With that said: the concept of validity range is *not* part of the application level protocol today, I don't see a reason to add this as it creates more problems than it solves (and a PR is not the place to discuss this anyways).

rokopt replied:
> With that said: the concept of validity range is _not_ part of the application level protocol today, I don't see a reason to add this as it creates more problems than it solves (and a PR is not the place to discuss this anyways).

I don't understand why you're saying that "validity range" isn't part of the protocol today. Today, a `StellarValue` `v` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
max(map(minTime,v.txs)) <= lcl.closeTime
min(map(maxTime,v.txs)) >= lcl.closeTime
```

In other words, `lcl.closeTime` already has a validity range, namely `[max(map(minTime,v.txs))]`.

After my change, a `StellarValue` `v` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
max(map(minTime,txs(v))) <= lcl.closeTime
min(map(maxTime,txs(v)) >= v.closeTime
```

I haven't changed the form of the inequalities; I've just made the second inequality stricter (it's definitely stricter as opposed to equivalent or weaker, because we also validate `v.closeTime > lcl.closeTime`).

Or, if you look at it from the perspective of the individual transactions rather than the `StellarValue`, then today, a given transaction `tx` is valid (as far as `closeTime` is concerned) if and only if the following hold:

```
minTime(tx) <= lcl.closeTime
maxTime(tx) >= lcl.closeTime
```

So `minTime(tx)` already had a validity range, namely `(-inf,lcl.closeTime])`, and `maxTime(tx)` already had a validity range, namely `[lcl.closeTime,+inf)`.

After my change, a given transaction `tx` is valid in a `StellarValue` `v` (as far as `closeTime` is concerned) if and only if the following hold:

```
minTime(tx) <= lcl.closeTime
maxTime(tx) >= v.closeTime
```

So in my code, `minTime(tx)`'s validity range remains the same, and `maxTime(tx)`'s validity gets smaller on the left, becoming `[v.closeTime,+inf)`.

How is it that validity ranges are not there already?  It looks to me as though they're already there, and I'll I've done is change the location of one of the boundaries (which I thought is exactly what CAP-34 said I should do).

rokopt replied:
Nicolas has explained that he _does_ want to give us the flexibility to relax the flooding as well in the future, so it would be best to weaken the test for `txTOO_EARLY` to the precise degree that we can, namely, against the `closeTime` in the `StellarValue` itself, like the test for `txTOO_LATE` -- though the latter is a strengthening.  So the overall change that we want is not (as I have been thinking) a strengthening of validation, but a strengthening of one test and a weakening of another, with the overall result being neither a weakening nor a strengthening, but a _shift_, of the whole mechanism of `minTime`/`maxTime`/`closeTime` validation from `lcl.closeTime` to `v.closeTime`.

rokopt replied:
(I'll make that semantic change after addressing other code review comments.)

rokopt replied:
(I'll also update the CAP to reflect that.)

rokopt replied:
Alright, I have made the big semantic change in commit 14a693a04985274b003047da09196554b82cb3bc, by extending the PR #2608 code to include a `lowerBoundCloseTimeOffset` as well as the `upperBoundCloseTimeOffset` which that PR added.  That allows my new CAP-34 code to use the `lowerBoundCloseTimeOffset` to specify which close time to validate `minTime` against, as it was already using `upperBoundCloseTimeOffset` to specify which close time to validate `maxTime` against.  (In the two key places where the semantics change -- `HerderImpl::triggerNextLedger()` and `HerderSCPDriver::validateValueHelper()` -- the CAP-34 code passes the same value for both offsets (the difference between the proposed next `closeTime` and the last ledger close time, meaning that both `minTime` and `maxTime` will be validated against the proposed next `closeTime`).)

MonsieurNicolas commented on src/herder/HerderImpl.cpp:
```diff
@@ -846,8 +846,25 @@ HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger)
+        nextCloseTime = lcl.header.scpValue.closeTime + 1;
+    }
+
+    // Protocols including the "closetime change" (CAP-0034) externalize
```
I don't think this is what CAP0034 says: we expect txsets to all be valid for the new close time.

This change implies that we expect them to be valid from `lcl.ct` up to `new_ct` (inclusive) which is a lot more restrictive.

rokopt replied:
(Same question as to the previous comment.)

rokopt replied:
(Same upcoming resolution.)

rokopt commented on src/herder/HerderImpl.cpp:
```diff
@@ -605,16 +605,17 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
-                                         return true;
-                                     },
-                                     false);
+        getSCP().processCurrentState(
```
My clang-format keeps insisting on that change; should I remove it anyway?  I'm running 5.0.2; is that not the right version?

rokopt replied:
(I've removed the whole commit now to put into another branch, so this question is moot.)

jonjove replied:
I don't think that is the right version. Unfortunately, even tiny changes to the version will potentially change the formatting (I remember that @MonsieurNicolas and I had different tags of the 5.0.0 release which caused issues). My clang-format reports
```
clang-format version 5.0.0 (tags/RELEASE_500/final)
```

rokopt replied:
> I remember that @MonsieurNicolas and I had different tags of the 5.0.0 release which caused issues

Sigh!  `clang-format` is the sort of thing that I would hope would provide backwards compatibility (given the same config file) even across major versions ("`--behave-like-version=5.0.0`" or something), let alone tags of the same release.

At any rate, I've moved that whole commit to a new PR, and in the new PR I've removed the changes that `clang-format` made.

rokopt commented on src/herder/test/HerderTests.cpp:
```diff
@@ -51,7 +51,6 @@ TEST_CASE("standalone", "[herder][acceptance]")
     cfg.QUORUM_SET.validators.push_back(v0NodeID);
 
     for_all_versions(cfg, [&](Config const& cfg1) {
-
```
My clang-format wants to make this one too -- is that not supposed to be the behavior here?

rokopt replied:
(I've removed the whole commit now to put into another branch, so this question is moot.)

rokopt commented on src/historywork/DownloadVerifyTxResultsWork.cpp:
```diff
@@ -67,9 +67,11 @@ DownloadVerifyTxResultsWork::yieldMoreWork()
-        mApp, "download-verify-results-" + std::to_string(mCurrCheckpoint),
-        seq);
+    auto w3 =
+        std::make_shared<WorkSequence>(mApp,
```
Alright.  I've moved that commit (with the `fmt::format()` change) to https://github.com/rokopt/stellar-core/tree/refactoring-2 .

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -18,13 +18,18 @@
 #include "util/Logging.h"
 #include "xdr/Stellar-SCP.h"
 #include "xdr/Stellar-ledger-entries.h"
+#include "xdr/Stellar-ledger.h"
```
I changed the commit messages to link to the CAP instead of the PR.

rokopt commented on src/transactions/test/TxEnvelopeTests.cpp:
```diff
@@ -1531,6 +1534,22 @@ TEST_CASE("txenvelope", "[tx][envelope]")
                         }
                     }
+
+                    SECTION("accepted on time, but closed later "
```
I've removed the redundant test.

I've also removed that whole commit, and made a new branch, and opened https://github.com/stellar/stellar-core/pull/2626 for it.

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -324,39 +342,29 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
+    // the close time to validate against is not the last ledger close
+    // time, but the proposed close time (which we have validated above,
+    // in checkCloseTime(), is greater than the last ledger close time).
+    if (b.closeTime <= lastCloseTime)
```
That's why I called it a `logic_error`; I wanted to make sure it'd be detected if something changed earlier in the method so that that wasn't guaranteed anymore.  I'll remove it.

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -697,11 +700,20 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     {
         Hash highest;
         TxSetFramePtr highestTxSet;
+        if (candidateValues.empty())
```
Moved (but `candidates` is the parameter, not `candidateValues`, which is a local, so at the top of the function it becomes `candidates` that we test).

rokopt commented on src/herder/HerderSCPDriver.h:
```diff
@@ -156,6 +157,31 @@ class HerderSCPDriver : public SCPDriver
+    // to validate or trim the transaction set in the StellarValue against?
+    // (The parameters and return value of this method are specified as
+    // _offsets_ from the last ledger close time.)
+    virtual TimePoint closeTimeOffsetToTrimOrValidateAgainst(
```
Alright, I removed the machinery.  Back to `if`s and ternary operators.  :)

MonsieurNicolas replied:
yeah sorry for the churn, but when it takes me more than 10 seconds to understand what a function does even *with* code navigation, I get nervous :)

rokopt commented:
By the way, I request that someone who knows what they're doing (i.e. not me, yet at least) take a particular look at this, which I called out in the log message:

```
The checking of signatures (or for their absence) in
HerderSCPDriver::validateValueHelper() had been split into
two different parts of the method, and is consolidated into
one place, at the beginning, by this change.
```

I don't understand why that logic, which checks whether the `SIGNED`/`BASIC` type is what we expect, and if `SIGNED`, whether the signature is correct, was in two separate, partly-but-not-completely-redundant checks in different parts of the method, instead of all together in one place at the beginning of the method, where I've consolidated it.  As far as I can tell, all the inputs to the checks -- `b`, `nomination`, and `lcl` -- are all constants, so the intervening "if this value is not for our local state" code shouldn't change the outcome of the signature-related checks.

But it does occur to me now that that intervening code does return `kMaybeValidValue` in a couple of places, even where the signature state might have been incorrect according to the first test, so that my new code would return `kInvalidValue` in those cases -- specifically, the old code didn't test the ledger version, so I think it could have returned `kMaybeValidValue` in some cases, in particular `if (!mTrackingSCP)`, if the ledger version were `< 11` and the value was `SIGNED`, whereas my new code would reject the `SIGNED` value because protocol versions `< 11` used `BASIC` values even in the nomination phase.  Am I breaking something, such as upgrades from version `10` to version `11` because we would have to accept some `SIGNED` value as `kMaybeInvalid` because, if we're "not tracking SCP", then maybe the other nodes have upgraded to version `11`?  I'm not sure precisely what "tracking SCP" means, so I'm not sure.

rokopt commented:
> Am I breaking something, such as upgrades from version `10` to version `11` because we would have to accept some `SIGNED` value as `kMaybeInvalid` because, if we're "not tracking SCP", then maybe the other nodes have upgraded to version `11`? I'm not sure precisely what "tracking SCP" means, so I'm not sure.

Upon even further reflection, I don't think I could have broken that, because if we're ever to upgrade, we'll need to receive a message that upgrades to version `11`, and that message would itself be using the version `10` protocol, and would therefore be `BASIC`.  So I still can't think of anything that I'm breaking.  But I don't know much.  :)

jonjove commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -351,7 +354,9 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
         res = SCPDriver::kInvalidValue;
     }
-    else if (!txSet->checkValid(mApp, 0))
+    else if (!txSet->checkValid(mApp, curProtocolContainsClosetimeChange()
```
This seems suspicious to me. Suppose you have a transaction set containing a single transaction which is valid after time `T`. For a value with close-time `T+1`, this transaction is valid. But if the last closed ledger has close-time `T-1`, I think this check will erroneously consider this transaction invalid.

This is closely related to the "validity range" comments from @MonsieurNicolas.

Is there a test case for this? If so, which one covers it? If not, can you add one?

rokopt replied:
By "valid after time `T`", you mean it has a `minTime` of `T`, correct?  If the last ledger has closetime `T-1`, that will indeed be considered invalid, specifically `txTOO_EARLY`.  And yes, I think I gather that that's what @MonsieurNicolas is objecting to as well.

What I don't understand is why that's erroneous.  That's exactly what we would have done before my change, isn't it?  And CAP-0034 only mentions `txTOO_LATE` and `maxTime`; it has zero mention of either `txTOO_LATE` or `minTime`.  And I don't understand why it _would_ mention `txTOO_EARLY` or `minTime`, because they do not suffer from the problems that `txTOO_LATE` and `maxTime` do which CAP-0034 is trying to address.

I also don't understand how a transaction set could even _reach_ this code if it contained a transaction with `minTime==T` in the presence of the condition `lcl==T-1` , because such a transaction would have been trimmed way back in `TransactionQueue::canAdd()`.  At that point, we don't _have_ any next-`closeTime` estimate, so the _only_ thing we can validate against is `lcl.closeTime`.  I therefore don't understand how adding a bunch of code to allow transactions with `minTime > lcl.closeTime` at this point would even change the system's semantics.

If there is actually some reason that it would make a difference to validate `minTime`s against the `closeTime` in the `StellarValue` at this point, then I not only have to write a bunch more code, but also a bunch more documentation, because I did not understand that the CAP that I myself was writing implied a change to the validation of `minTime` as well as that of `maxTime`.  (There's no occurrence of either `minTime` or `txTOO_EARLY` in the discussion in the entire discussion in the [issue](https://github.com/stellar/stellar-protocol/issues/622), either.)

jonjove replied:
It is not exactly what we would have done before. Here is an example. Suppose you have a transaction set containing a single transaction which is valid after time `T` and before time `T` (I believe both the time bounds are inclusive, but if that's not true then just assume it for the sake of this example). For a value with close-time `T`, this transaction should be valid. But with your approach it is _impossible_ for this transaction set to be valid, because the close-time of two consecutive ledgers must differ by at least one second. In the existing protocol, this transaction set is at least valid if the previous close-time was exactly `T`.

Furthermore, as @MonsieurNicolas noted, you are conflating implementation details with protocol details. What stops me from modifying stellar-core to accept transactions that I anticipate will be valid at the next ledger close? Nothing, because that is an implementation detail. But once I have done that, I could easily include such transactions in a transaction set that I nominate. Because I am nominating a <close-time, txset> pair, such a transaction set should be valid if every transaction in the set would be valid at that close-time.

The key point here is that what we were doing before is wrong. That's the whole reason that CAP-34 exists. Just because your proposed change makes the situation better than it was before, doesn't mean it's the best possible change. It is the duty of the author to evaluate all of the potential consequences of the proposal, and thereby achieve the best properties that are reasonable. From my CAP-15 blog post https://www.stellar.org/developers-blog/how-we-simplified-fees-on-the-stellar-network:

> First, even changes that sound simple, such as making it possible to increase the fee on a transaction, can be extremely challenging to design and implement. Second, we need to strike the right balance between providing the strongest guarantees that we can and designing features which can actually be implemented safely.

In this case, I think providing stronger guarantees will make the implementation easier to reason about and therefore safer.

rokopt replied:
> It is not exactly what we would have done before. Here is an example. Suppose you have a transaction set containing a single transaction which is valid after time `T` and before time `T` (I believe both the time bounds are inclusive, but if that's not true then just assume it for the sake of this example). For a value with close-time `T`, this transaction should be valid. But with your approach it is _impossible_ for this transaction set to be valid, because the close-time of two consecutive ledgers must differ by at least one second. In the existing protocol, this transaction set is at least valid if the previous close-time was exactly `T`.

I think I understand you to be saying that the transaction has `minTime==T` and `maxTime==T`, and the `StellarValue` `v` that the transaction is in has `v.closeTime==T`.  Correct?

But the old code here validated `minTime` and `maxTime` against `lcl.closeTime`, and with `minTime==maxTime`, `lcl.closeTime==minTime==maxTime` is the only way to avoid both `txTOO_EARLY` and `txTOO_LATE`.  Correct?

And as you point out, for a `StellarValue` `v` to be valid, it must also have `v.closeTime > lcl.closeTime` (strict inequality).  But we have also shown that validity requires `lcl.closeTime==minTime==T==v.closeTime`.  This is a contradiction.  So the old code would also have failed the validity check here, specifically when it tried to validate that `v.closeTime` was strictly greater than `lcl.closeTime`.  Correct?

> Furthermore, as @MonsieurNicolas noted, you are conflating implementation details with protocol details. What stops me from modifying stellar-core to accept transactions that I anticipate will be valid at the next ledger close? Nothing, because that is an implementation detail. But once I have done that, I could easily include such transactions in a transaction set that I nominate. Because I am nominating a <close-time, txset> pair, such a transaction set should be valid if every transaction in the set would be valid at that close-time.

The protocol also doesn't require that we _do_ accept transactions that weren't valid at the last ledger close, does it?  In particular, the existing code in _our_ stellar-core would have rejected such a transaction, so it seems to me that that surely wouldn't be a violation of the protocol.  In other words, for some other implementation to accept transactions that it _anticipates_ to be valid (i.e. not too early) at the time of the next ledger close is that implementation's free choice of implementation detail, and for our implementation _not_ to accept such a transaction unless it was already valid (i.e. not too early) at the time of last ledger close is _our_ free choice of implementation detail.

> The key point here is that what we were doing before is wrong. That's the whole reason that CAP-34 exists. Just because your proposed change makes the situation better than it was before, doesn't mean it's the best possible change. It is the duty of the author to evaluate all of the potential consequences of the proposal, and thereby achieve the best properties that are reasonable. From my CAP-15 blog post https://www.stellar.org/developers-blog/how-we-simplified-fees-on-the-stellar-network:
> 
> > First, even changes that sound simple, such as making it possible to increase the fee on a transaction, can be extremely challenging to design and implement. Second, we need to strike the right balance between providing the strongest guarantees that we can and designing features which can actually be implemented safely.
> 
> In this case, I think providing stronger guarantees will make the implementation easier to reason about and therefore safer.

I agree -- and I'm arguing for a stronger guarantee.  Changing my code to validate `minTime` against `v.closeTime` instead of `lcl.closeTime`, as I gather you and Nicolas are suggesting, would make the guarantees that we make about valid `StellarValue`s strictly weaker (wouldn't it?).

rokopt replied:
See [this comment](https://github.com/stellar/stellar-core/pull/2625#discussion_r458379008) for the upcoming resolution.

jonjove commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -619,6 +624,11 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     CLOG(DEBUG, "Herder") << "Combining " << candidates.size() << " candidates";
     mSCPMetrics.mCombinedCandidates.Mark(candidates.size());
 
+    if (candidates.empty())
```
Why is this needed now, when it wasn't before?

rokopt replied:
It's not needed; it's an assertion, which I thought, if a caller of `combinedCandidates()` were ever to turn out to have a bug that called that method with an empty `candidates` list, would make the source clearer than the `NULL` pointer dereference that we'd get without it.  I'll remove it.

jonjove commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -692,6 +703,9 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     TxSetFramePtr bestTxSet;
+    TimePoint bestTxSetCloseTime = 0;
+    LedgerCloseValueSignature highestTxSetSignature;
+    auto const newCloseTimeBehavior = curProtocolContainsClosetimeChange();
```
Nitpick: same thing as for the global version constant. This name doesn't tell me anything about what that behavior is, or even what "new" means. I can understand it just fine in the context of this PR, but when someone maintains this code in a year or two this name won't mean anything to them.

rokopt replied:
I've changed it to `curProtocolPreservesTxSetCloseTimeAffinity()`, matching the CAP title (and thus the new constant name as well).

jonjove commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -692,6 +703,9 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
 
     // take the txSet with the biggest size, highest xored hash that we have
     TxSetFramePtr bestTxSet;
+    TimePoint bestTxSetCloseTime = 0;
```
Relating to what @marta-lokhova commented below, I don't think you actually need this at all? Or am I missing something?

rokopt replied:
If you mean that I could replace `bestTxSetTime = sv.closeTime` with `comp.closeTime = sv.closeTime` (under `newCloseTimeBehavior`), then yeah, I could do that.  I thought it was clearer to have the decision point based on whether we're using pre- or post-CAP-34 code in one place, namely here:

```
    if (newCloseTimeBehavior)
    {
        comp.closeTime = bestTxSetCloseTime;
        comp.ext.v(STELLAR_VALUE_SIGNED);
        comp.ext.lcValueSignature() = highestTxSetSignature;
    }
    else
    {
        comp.closeTime = maxCloseTime;
        comp.ext.v(STELLAR_VALUE_BASIC);
    }
```

I thought that that spelled out in one place exactly what the difference was between the way that the old and new protocols constructed `comp`, whereas if I didn't use the locals, then the assignment for 'newCloseTimeBehavior' would be up in the loop (and possibly done repeatedly) whereas the one for `!newCloseTimeBehavior` would be after the loop, making it less immediately obvious what the difference was, and perhaps making it unclear why we were not doing a post-loop assignment after all in the `newCloseTime` case.  It appears not to be providing that clarity, so I'll change it.

jonjove commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -845,6 +858,13 @@ HerderSCPDriver::valueExternalized(uint64_t slotIndex, Value const& value)
 }
 
+bool
+HerderSCPDriver::curProtocolContainsClosetimeChange() const
```
Nitpick: name

rokopt replied:
Changed as noted above.

jonjove commented on src/herder/test/HerderTests.cpp:
```diff
@@ -1066,8 +1072,12 @@ TEST_CASE("surge pricing", "[herder][txset]")
+              bool checkHighFee, bool withSCPsignature,
+              bool const expectNewClosetimeBehavior)
 {
+    using CloseTime = decltype(StellarValue::closeTime);
```
This doesn't add clarity, because you are using names different from our convention. When I read `CloseTime`, my mind questions what type that is rather than automatically jumping to `TimePoint`. Is there a good reason not to use `TimePoint` and `UpgradeType`, like we do everywhere else?

rokopt replied:
`StellarValue::upgrades` isn't of type `UpgradeType`; it's of type `xdr::xvector<UpgradeType,6>`, which made me a little queasy to write in a bunch of places (hence the `using`) -- or even one, for that matter (hence the `decltype`).  After that I just felt strange treating the `closeTime` and `upgrades` fields differently as far as whether I was inferring their types or not.

I've gotten rid of the `CloseTime` in favor of `TimePoint`.  Do you definitely also want me to replace `SVUpgrades` with `xdr::xvector<UpgradeType,6>`?

rokopt commented on src/herder/HerderSCPDriver.h:
```diff
@@ -156,6 +157,18 @@ class HerderSCPDriver : public SCPDriver
 
+    // Does the nomination protocol output a BASIC or a SIGNED
+    // StellarValue?
+    virtual StellarValueType typeProducedByNomination() const;
```
Done.

rokopt commented on src/herder/HerderSCPDriver.h:
```diff
@@ -4,6 +4,7 @@
 // under the Apache License, Version 2.0. See the COPYING file at the root
 // of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
 
+#include "fmt/format.h"
```
Removed.

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -203,24 +208,42 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
     uint64_t lastCloseTime;
     ZoneScoped;
+    auto const& lcl = mLedgerManager.getLastClosedLedgerHeader().header;
+    bool const expectSignedValue = nomination
```
Thanks for catching, and explaining, that.  I have reverted the original two signature-checking-related changes in HerderSCPDriver::validateValueHelper() and have tried again in light of this.  Would you please let me know whether it looks right now?

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -706,6 +720,16 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
+                    bestTxSetCloseTime = sv.closeTime;
+                    if (newCloseTimeBehavior)
+                    {
+                        if (sv.ext.v() != STELLAR_VALUE_SIGNED)
```
> Isn't this already checked by validateValueHelper?

Yeah, I just meant it as an assertion that that method had done what combineCandidates() expects.  If this kind of assertion is more confusing than clarifying, I will try to start avoiding them.

> Also, you should be able to simply use the candidate comp = sv like it's specified in the proposal instead of manually crafting a new value like this

We're not taking `sv.upgrades`, though, are we?  That's still a combination of all candidates' upgrades.  Are you saying it would be clearer to do `comp = sv` and then overwrite the `upgrades`?  I thought that might have been misleading (back at the location of the `comp = sv`).

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -18,13 +18,18 @@
 namespace stellar
 {
+uint32_t const
+    HerderSCPDriver::FIRST_PROTOCOL_VERSION_CONTAINING_CLOSETIME_CHANGE = 14;
```
I changed it to FIRST_PROTOCOL_WITH_TXSET_CLOSETIME_AFFINITY, directly echoing the CAP.

rokopt commented:
> By the way, I request that someone who knows what they're doing (i.e. not me, yet at least) take a particular look at this, which I called out in the log message:
> 
> ```
> The checking of signatures (or for their absence) in
> HerderSCPDriver::validateValueHelper() had been split into
> two different parts of the method, and is consolidated into
> one place, at the beginning, by this change.
> ```
> 
> I don't understand why that logic, which checks whether the `SIGNED`/`BASIC` type is what we expect, and if `SIGNED`, whether the signature is correct, was in two separate, partly-but-not-completely-redundant checks in different parts of the method, instead of all together in one place at the beginning of the method, where I've consolidated it. As far as I can tell, all the inputs to the checks -- `b`, `nomination`, and `lcl` -- are all constants, so the intervening "if this value is not for our local state" code shouldn't change the outcome of the signature-related checks.
> 
> But it does occur to me now that that intervening code does return `kMaybeValidValue` in a couple of places, even where the signature state might have been incorrect according to the first test, so that my new code would return `kInvalidValue` in those cases -- specifically, the old code didn't test the ledger version, so I think it could have returned `kMaybeValidValue` in some cases, in particular `if (!mTrackingSCP)`, if the ledger version were `< 11` and the value was `SIGNED`, whereas my new code would reject the `SIGNED` value because protocol versions `< 11` used `BASIC` values even in the nomination phase. Am I breaking something, such as upgrades from version `10` to version `11` because we would have to accept some `SIGNED` value as `kMaybeInvalid` because, if we're "not tracking SCP", then maybe the other nodes have upgraded to version `11`? I'm not sure precisely what "tracking SCP" means, so I'm not sure.

Note:  Marta answered this in [a review comment](https://github.com/stellar/stellar-core/pull/2625#discussion_r455438706) (and a change did turn out to be necessary).

rokopt commented:
I've now addressed all the review comments made so far, as far as I can tell.  In particular, I made the big semantic change to validate `minTime` (for `txTOO_EARLY`) as well as `maxTime` (for `txTOO_LATE`) against the next proposed `closeTime`, in the places where we have a proposed `closeTime`.  I've also rebased and squashed commits.  So I believe that this is ready for the next round of reviews.

rokopt commented:
> In particular, I made the big semantic change to validate minTime (for txTOO_EARLY) as well as maxTime (for txTOO_LATE) against the next proposed closeTime, in the places where we have a proposed closeTime.

That unfortunately made the diffs much larger, but I think it at least should facilitate this plan:

> Nicolas has explained that he does want to give us the flexibility to relax the flooding as well in the future

The `checkValid()`, `checkOrTrim()`, and `trimInvalid()` methods (among others) get `lowerBoundCloseTimeOffset` parameters to go with their `upperBoundCloseTimeOffset` parameters from PR #2608 , which should allow `HerderImpl::triggerNextLedger()`, `HerderImpl::updateTransactionQueue()`, and `TransactionQueue::canAdd()` to call some new `getLowerBoundCloseTImeOffset()` heuristic method to pass in as the new `lowerBoundCloseTimeOffset` parameter to `trimInvalid()`/`checkValid()`.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -319,18 +316,17 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
 
-    if ((!nomination || lcl.ledgerVersion < 11) &&
-        b.ext.v() != STELLAR_VALUE_BASIC)
+    bool const expectSignedValue =
```
let's add a commit before the CAP34 changes that *removes* the pre protocol 11 behavior entirely, that way we can have this commit focus on new behavior only

rokopt replied:
> let's add a commit before the CAP34 changes that _removes_ the pre protocol 11 behavior entirely, that way we can have this commit focus on new behavior only

Will that induce any change in this commit besides allowing...

```
    bool const expectSignedValue =
        nomination ? (lcl.ledgerVersion >= 11)
                   : (compositeValueType() == STELLAR_VALUE_SIGNED);
```

...to be reduced to...

```
    bool const expectSignedValue =
        nomination || (compositeValueType() == STELLAR_VALUE_SIGNED);
```

...?  (We could also remove that `SECTION("protocol 10")` test that you mentioned below, but that was there already.)

What will removing pre-protocol-11 behavior entail?  I see at least dozens of comparisons of ledger versions `> 10` or `>= 11`, as well as with lower version numbers, throughout the code base; are you talking about removing all of that, and starting to ignore (?) messages from nodes speaking protocols lower than 11?

rokopt replied:
Nico explained to me what he meant by removing pre-protocol-11 behavior, and I have now done that in a commit prior to the main CAP-34 commit.

MonsieurNicolas commented on src/herder/HerderImpl.cpp:
```diff
@@ -845,17 +845,40 @@ HerderImpl::triggerNextLedger(uint32_t ledgerSeqToTrigger)
+    // transaction set, so we know the exact closeTime against which to
+    // validate here -- 'nextCloseTime'.  (The _offset_, therefore, is
+    // the difference between 'nextCloseTime' and the last ledger close time.)
+    auto upperBoundCloseTimeOffset =
```
As those behaviors are supposed to work in lockstep, it would be better to write the code like this:
```
    if (getHerderSCPDriver().curProtocolPreservesTxSetCloseTimeAffinity())
    {
        upperBoundCloseTimeOffset = nextCloseTime - lcl.header.scpValue.closeTime;
        lowerBoundCloseTimeOffset = upperBoundCloseTimeOffset;
    }
    else
    {
        upperBoundCloseTimeOffset = getUpperBoundCloseTimeOffset(mApp, lcl.header.scpValue.closeTime);
        lowerBoundCloseTimeOffset = 0;
    }
```

rokopt replied:
Agreed and changed.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -319,18 +316,17 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
-    if ((!nomination || lcl.ledgerVersion < 11) &&
-        b.ext.v() != STELLAR_VALUE_BASIC)
+    bool const expectSignedValue =
+        nomination ? (lcl.ledgerVersion >= 11)
```
In this commit, you're missing enforcement that we have signed values all the time after protocol 14 (if `lcl.ledgerVersion >= 14` we expect everything to be signed).

I am not sure why some things are checked here and not at the top of the file (there is no point in returning "maybe valid" if there is no way values are valid because they're not signed)

rokopt replied:
> In this commit, you're missing enforcement that we have signed values all the time after protocol 14 (if `lcl.ledgerVersion >= 14` we expect everything to be signed).

Why is that?  If `nomination` is `true`, then for protocols after `14`, `expectSignedValue` will be `true` because it's `true` for all protocols after `11`, and if `nomination` is `false` (i.e. we're in the ballot protocol), then for protocols after `14`, `expectedSignedValue` will be `true` because `compositeValueType()` will return `STELLAR_VALUE_SIGNED` for all such protocols.

> I am not sure why some things are checked here and not at the top of the file (there is no point in returning "maybe valid" if there is no way values are valid because they're not signed)

Is that not one of the cases that Marta was referring to in [this comment](https://github.com/stellar/stellar-core/pull/2625#discussion_r455438706)?

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -351,7 +347,13 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
     }
-    else if (!txSet->checkValid(mApp, 0))
+    else if (!txSet->checkValid(mApp,
+                                curProtocolPreservesTxSetCloseTimeAffinity()
```
those two parameters are equal, we should use a variable instead of duplicating

rokopt replied:
Agreed and changed.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -706,6 +711,12 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
                 {
                     highestTxSet = cTxSet;
                     highest = sv.txSetHash;
+                    if (preserveTxSetCloseTimeAffinity)
```
I think we can move this out of the loop and use iterators like I suggested before. It avoids copies, and I think the code is easier to read:
```c++
diff --git a/src/herder/HerderSCPDriver.cpp b/src/herder/HerderSCPDriver.cpp
index 2991f200..b0cc2d08 100644
--- a/src/herder/HerderSCPDriver.cpp
+++ b/src/herder/HerderSCPDriver.cpp
@@ -694,50 +694,43 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     }

     // take the txSet with the biggest size, highest xored hash that we have
-    TxSetFramePtr bestTxSet;
     auto const preserveTxSetCloseTimeAffinity =
         curProtocolPreservesTxSetCloseTimeAffinity();
+    auto highest = candidateValues.end();
     {
-        Hash highest;
         TxSetFramePtr highestTxSet;
-        for (auto const& sv : candidateValues)
+        for (auto it = candidateValues.begin(); it != candidateValues.end();
+             ++it)
         {
+            auto const& sv = *it;
             TxSetFramePtr cTxSet = mPendingEnvelopes.getTxSet(sv.txSetHash);

             if (cTxSet && cTxSet->previousLedgerHash() == lcl.hash)
             {
-                if (compareTxSets(highestTxSet, cTxSet, highest, sv.txSetHash,
-                                  lcl.header, candidatesHash))
+                if (highest == candidateValues.end() ||
+                    compareTxSets(highestTxSet, cTxSet, highest->txSetHash,
+                                  sv.txSetHash, lcl.header, candidatesHash))
                 {
                     highestTxSet = cTxSet;
-                    highest = sv.txSetHash;
-                    if (preserveTxSetCloseTimeAffinity)
-                    {
-                        comp.closeTime = sv.closeTime;
-                        comp.ext.v(STELLAR_VALUE_SIGNED);
-                        comp.ext.lcValueSignature() = sv.ext.lcValueSignature();
-                    }
+                    highest = it;
                 }
             }
         }
-        // make a copy as we're about to modify it and we don't want to mess
-        // with the txSet cache
-        bestTxSet = std::make_shared<TxSetFrame>(*highestTxSet);
     }

+    comp = *highest;
     if (!preserveTxSetCloseTimeAffinity)
     {
         comp.closeTime = maxCloseTime;
         comp.ext.v(STELLAR_VALUE_BASIC);
     }
+    comp.upgrades.clear();
     for (auto const& upgrade : upgrades)
     {
         Value v(xdr::xdr_to_opaque(upgrade.second));
         comp.upgrades.emplace_back(v.begin(), v.end());
     }

-    comp.txSetHash = bestTxSet->getContentsHash();
-
     auto res = wrapStellarValue(comp);
     return res;
 }
```

rokopt replied:
Okay, I like that too.  Adopted.

rokopt replied:
In fact, as long as we're using iterators, why not go all the way and use the STL?  No explicit copying in the loop, and indeed no explicit loop, just a functional expression that we're finding the maximum of the list with an inline comparison function.

```
    // take the txSet with the biggest size, highest xored hash that we have
    using HighestPair = std::pair<StellarValue, TxSetFramePtr>;
    HighestPair highestPair = std::accumulate(
        candidateValues.begin(), candidateValues.end(), HighestPair(),
        [&](HighestPair const& hp, StellarValue const& sv) {
            auto const& highestTxSet = hp.second;
            auto const& svSetHash = sv.txSetHash;
            auto cTxSet = mPendingEnvelopes.getTxSet(svSetHash);
            return (cTxSet && cTxSet->previousLedgerHash() == lcl.hash &&
                    (!highestTxSet ||
                     compareTxSets(highestTxSet, cTxSet, hp.first.txSetHash,
                                   svSetHash, lcl.header, candidatesHash)))
                       ? HighestPair(sv, cTxSet)
                       : hp;
        });

    comp = highestPair.first;
    if (!curProtocolPreservesTxSetCloseTimeAffinity())
    {
        comp.closeTime = maxCloseTime;
        comp.ext.v(STELLAR_VALUE_BASIC);
    }
    comp.upgrades.clear();
```

MonsieurNicolas commented on src/herder/test/HerderTests.cpp:
```diff
@@ -1451,12 +1613,16 @@ TEST_CASE("SCP Driver", "[herder][acceptance]")
     SECTION("protocol 10")
     {
-        testSCPDriver(10, 10, 10, false, false);
+        testSCPDriver(10, 10, 10, false, false, false);
```
let's remove the behavior pre protocol 11 in a previous commit

rokopt replied:
Done.

rokopt commented on src/herder/test/PendingEnvelopesTests.cpp:
```diff
@@ -35,10 +36,15 @@ TEST_CASE("PendingEnvelopes recvSCPEnvelope", "[herder]")
     auto a1 = TestAccount{*app, getAccount("A")};
     using TxPair = std::pair<Value, TxSetFramePtr>;
-    auto makeTxPair = [](TxSetFramePtr txSet, uint64_t closeTime) {
+    auto makeTxPair = [&s, &herder](TxSetFramePtr txSet, uint64_t closeTime) {
```
Changed (in two places).

rokopt commented:
I've committed and squashed the coding-style changes that Nico requested.  I don't understand either why the `expectSignedValue` test is wrong or what removing protocols below 11 would mean (or why that would help), so I'll need clarification before addressing those two remaining comments.

rokopt commented:
I realized that I should probably make the `HerderTests`' `externalize()` utility function sign the value that it produces, for consistency with what the production code will do:

```
@@ -2045,7 +2211,7 @@ TEST_CASE("In quorum filtering", "[quorum][herder][acceptance]")
 }

 static void
-externalize(LedgerManager& lm, HerderImpl& herder,
+externalize(SecretKey const& sk, LedgerManager& lm, HerderImpl& herder,
             std::vector<TransactionFrameBasePtr> const& txs)
 {
     auto const& lcl = lm.getLastClosedLedgerHeader();
@@ -2061,6 +2227,11 @@ externalize(LedgerManager& lm, HerderImpl& herder,

     StellarValue sv{txSet->getContentsHash(), 2, xdr::xvector<UpgradeType, 6>{},
                     STELLAR_VALUE_BASIC};
+    if (herder.getHerderSCPDriver().compositeValueType() ==
+        STELLAR_VALUE_SIGNED)
+    {
+        herder.signStellarValue(sk, sv);
+    }
     herder.getHerderSCPDriver().valueExternalized(ledgerSeq,
                                                   xdr::xdr_to_opaque(sv));
 }
```

rokopt commented:
Pre-protocol-11 behavior has been removed, and I think the PR is ready for re-review.

rokopt commented:
It also occurred to me that I wasn't thrilled at having added all those `, 0`s to calls to `checkValid()` and `trimInvalid()`, or at the thought of anyone having to do so again if any more parameters are ever added to those interfaces, so I've now given default values to both the new `lowerBoundCloseTimeOffset` and the pre-existing `upperBoundCloseTimeOffset` parameters.

MonsieurNicolas commented:
> It also occurred to me that I wasn't thrilled at having added all those , 0s to calls to checkValid() and trimInvalid(), or at the thought of anyone having to do so again if any more parameters are ever added to those interfaces, so I've now given default values to both the new lowerBoundCloseTimeOffset and the pre-existing upperBoundCloseTimeOffset parameters.

this was actually done on purpose in the PR that introduced the high offset: we want those to be explicit as to guarantee that we never ever pass in the wrong values for this (which would be catastrophic).

rokopt commented:
> > It also occurred to me that I wasn't thrilled at having added all those , 0s to calls to checkValid() and trimInvalid(), or at the thought of anyone having to do so again if any more parameters are ever added to those interfaces, so I've now given default values to both the new lowerBoundCloseTimeOffset and the pre-existing upperBoundCloseTimeOffset parameters.
> 
> this was actually done on purpose in the PR that introduced the high offset: we want those to be explicit as to guarantee that we never ever pass in the wrong values for this (which would be catastrophic).

Aha, I see.  I've removed that last commit.

rokopt commented on src/transactions/TransactionFrame.cpp:
```diff
@@ -470,13 +473,15 @@ TransactionFrame::commonValid(SignatureChecker& signatureChecker,
                               AbstractLedgerTxn& ltxOuter,
                               SequenceNumber current, bool applying,
                               bool chargeFee,
+                              uint64_t lowerBoundCloseTimeOffset,
```
Yeah, I like that idea; I've added it.  Actually, I've added something even stronger -- if we're `applying`, then I think that `lowerBoundCloseTimeOffset` and `upperBoundCloseTimeOffset` must be not only equal to each other, but specifically equal to `0`:  when we're closing a ledger and applying transactions, we should be validating the transactions against the exact ledger header that's being closed (which, by the time we apply transactions, has been made into the "current" ledger header).

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -687,59 +691,34 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
         comp.upgrades.emplace_back(v.begin(), v.end());
     }
 
-    // just to be sure
```
I had preserved them in the code prior to reviews, but @MonsieurNicolas argued that they were confusing in the presence of CAP-0034:

https://github.com/stellar/stellar-core/pull/2625#discussion_r454748474

I have no objection to their being either there or not there, so I'd adapt the code to either of you convincing the other of a preference.  ;)

rokopt commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -687,59 +691,34 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
-        }
-        // make a copy as we're about to modify it and we don't want to mess
-        // with the txSet cache
-        bestTxSet = std::make_shared<TxSetFrame>(*highestTxSet);
```
Correct -- the "modify it" that the old comment referred to was the `trimInvalid()` section, which I've removed as [noted above](https://github.com/stellar/stellar-core/pull/2625#discussion_r461229945).

It occurs to me in light of your question, though, that I could fling a bunch of `const`s around to make it clear (and compiler-checked) that we're _not_ modifying anything (and therefore don't need to copy anything).  I've done so in an additional commit (554b79ff4059442b8d2de3fb4128c2f91041e10b) just in case anyone thinks that that's not a good idea, in which case I could easily remove it again.

MonsieurNicolas commented on src/herder/test/HerderTests.cpp:
```diff
@@ -1451,7 +1451,7 @@ TEST_CASE("SCP Driver", "[herder][acceptance]")
     SECTION("protocol 10")
     {
-        testSCPDriver(10, 10, 10, false, false);
+        testSCPDriver(10, 10, 10, false, true);
```
this section should be removed (in the context of this commit)

rokopt replied:
I've removed it.  That also made the `checkHighFee` parameter obsolete (all remaining calls to `testSCPDriver()` used `true`), so I removed that too.

MonsieurNicolas commented on src/herder/test/UpgradesTests.cpp:
```diff
@@ -1949,7 +1949,7 @@ TEST_CASE("simulate upgrades", "[herder][upgrades][acceptance]")
         auto checks = std::vector<LedgerUpgradeCheck>{
             {genesis(0, 10), {noUpgrade, noUpgrade, noUpgrade}},
-            {genesis(0, 21), {upgrade, upgrade, upgrade}}};
+            {genesis(0, 28), {upgrade, upgrade, upgrade}}};
```
what are those changes?

rokopt replied:
The ones I described in this commit log:

> This required the timeouts on three upgrade tests to be
increased (because, I believe, of the extra signature-
checking during the nomination phase while running protocol
0).

MonsieurNicolas replied:
this doesn't make sense to me:
* we're in virtual time, so signature verification time is free
* in general, as the new behavior is supposed to pick (potentially) smaller close times, the ledger close time should maybe move in the other direction
* for this particular test, all nodes vote exactly the same than before, so I would expect the close times to be the same

I suggest you analyze what is really going on here...

rokopt replied:
> in general, as the new behavior is supposed to pick (potentially) smaller close times, the ledger close time should maybe move in the other direction

I'll certainly look into this, but it has nothing to do with the new closetime behavior -- it's the pre-commit to remove support for unsigned nominations (as they were in protocols 10 and before) that induces this test timeout, without the closetime change being present yet.

rokopt replied:
> we're in virtual time, so signature verification time is free

Ah, I figured the simulation must be charging us some time for verifying signatures.  If it doesn't, then yes, I am mystified.

rokopt replied:
I think I understand what's happening here now.  I think that when we made nominations signed back in protocol 11, we may have increased the number of nomination messages.

The reason is that `NominationProtocol::processEnvelope()` tests whether a vote can be accepted using voted and accepted `StatementPredicate`s (parameters to `Slot::federatedAccept()`) which test whether a nomination is already present in our `st.pledges.nominate().votes` or `st.pledges.nominate().accepted` vectors by doing a std::find, which tests for strict equality.

Prior to the signing of nominations, therefore, we would treat a nomination as if we had already voted for (resp. accepted) it if we had already voted for (resp. accepted) a nomination with the same `txSetHash`, `closeTime`, and `upgrades`.

But with the signing of nominations, we will only treat a nomination as if we had already voted for (resp. accepted) it if we had already voted for (resp. accepted) a nomination which not only had the same `txSetHash`, `closeTime`, and `upgrades` , but also originated with the same node (and therefore had the same signature).

So it ends up taking more messages to reach consensus.

I have no idea whether that difference is significant in production (for all I know, it might be extremely rare in real life that different nodes would generate the same nominations), but it certainly affects the `UpgradesTests`, which like to set up several nodes that nominate the same stuff.

To confirm that that's what causes the timeouts, I wrote the following code, and it made them go away:

https://github.com/rokopt/stellar-core/commit/9bbeb4e04859a925779b4b219816d681a508efb8

However, I doubt that that's actually correct; it's just a test of the hypothesis.  It makes some other tests fail -- in particular the SCP tests that count `mEnvs` start getting the wrong counts.  I'm not sure whether we actually want to change anything there or not, and if we do want to do a node-insensitive comparison, I'm not sure exactly what we should do when we find that we've voted for/accepted a nomination that is equal except that it comes from a different node. 

The upshot as far as CAP-34 is concerned, though:  the root cause of these test failures is the introduction of signed nominations in protocol 11, not any of the code that's been changed in this PR.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -687,59 +691,34 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
+            auto const& highestTxSet = hp.second;
+            auto const& svSetHash = sv.txSetHash;
+            auto const cTxSet = mPendingEnvelopes.getTxSet(svSetHash);
+            return (cTxSet && cTxSet->previousLedgerHash() == lcl.hash &&
```
The expression here though is quite hard to read (the old code was written like that as to make it straightforward to understand), so maybe preserving the old format helps here?

rokopt replied:
Alright, I removed the factoring-out of the repeatedly-used dereferences (as a result of removing the `std::accumulate` altogether; see below).

I kept the `if` statement squashed, though.  I don't see the nested `if`s as easier to read than one `if` with an `&&`.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -687,59 +691,34 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
     // take the txSet with the biggest size, highest xored hash that we have
-    TxSetFramePtr bestTxSet;
+    using HighestPair = std::pair<StellarValue, TxSetFramePtr>;
+    HighestPair highestPair = std::accumulate(
```
using the std library seems fine here... but it looks like it's not a home run:
* `std::accumulate` doesn't seem to be a great fit though as it copies the value all the time (which contains a hash). There is a version that uses `std::move` but that's C++ 20.
* there is `std::reduce` but that's C++17
* from what I see the only thing we can do is a straight loop at this time 🤷

rokopt replied:
Alright, I gave up on the STL.  I kept everything except the two accumulators (the pointers to the highest value and set) `const`, though (including the iterator, i.e. `cbegin()` and `cend()`).

MonsieurNicolas commented on src/herder/test/HerderTests.cpp:
```diff
@@ -1613,16 +1607,16 @@ TEST_CASE("SCP Driver", "[herder][acceptance]")
     SECTION("protocol 10")
     {
-        testSCPDriver(10, 10, 10, false, true, false);
+        testSCPDriver(10, 10, 10, false, false);
```
this test should have been removed in the previous commit

rokopt replied:
Removed.

rokopt commented:
I think I'm back to having addressed all code review comments.

MonsieurNicolas commented on src/herder/HerderSCPDriver.cpp:
```diff
@@ -691,59 +691,37 @@ HerderSCPDriver::combineCandidates(uint64_t slotIndex,
-        // with the txSet cache
-        bestTxSet = std::make_shared<TxSetFrame>(*highestTxSet);
+        };
+        comp = *highest;
```
out of paranoia, we should throw (`runtime_error`) if `highest == end`

rokopt replied:
Happy to!  Done.

rokopt replied:
> I suggest you analyze what is really going on here...

See https://github.com/stellar/stellar-core/pull/2625#discussion_r463089695 .

rokopt commented:
I think I'm once again back to having addressed all code review comments.

MonsieurNicolas commented:
r+ 2abee78c68a9028da1158a012e88d8477bcb94c3
